blocks/filter.py

```diff
@@ -0,0 +1,115 @@
+from inspect import isclass
+
+from blocks.bricks.base import ApplicationCall, Brick
+
+
+def get_annotation(var, cls):
+    """A helper function to retrieve an annotation of a particular type.
+
+    Notes
+    -----
+    This function returns the first annotation of a particular type. If
+    there are multiple--there shouldn't be--it will ignore them.
+
+    """
+    for annotation in getattr(var.tag, 'annotations', []):
+        if isinstance(annotation, cls):
+            return annotation
+
+
+def get_brick(var):
+    """Retrieves the brick that created this variable.
+
+    See :func:`get_annotation`.
+
+    """
+    return get_annotation(var, Brick)
+
+
+def get_application_call(var):
+    """Retrieves the application call that created this variable.
+
+    See :func:`get_annotation`.
+
+    """
+    return get_annotation(var, ApplicationCall)
+
+
+class VariableFilter(object):
+    """Filters Theano variables based on a range of criteria.
+
+    Parameters
+    ----------
+    roles : list of :class:`VariableRole` instances, optional
+        Matches any attribute which has one of the roles given.
```

Comment:
Yes, missed that one
On Jan 25, 2015 5:36 AM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/filter.py
> <https://github.com/bartvm/blocks/pull/141#discussion_r23503440>:
>
> > +def get_application_call(var):
> > +    """Retrieves the application call that created this variable.
> > +
> > +    See :func:`get_annotation`.
> > +
> > +    """
> > +    return get_annotation(var, ApplicationCall)
> > +
> > +
> > +class VariableFilter(object):
> > +    """Filters Theano variables based on a range of criteria.
> > +
> > +    Parameters
> > +    ----------
> > +    roles : list of :class:`VariableRole` instances, optional
> > +        Matches any attribute which has one of the roles given.
>
> Did you want to write "variable" instead of "attribute"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/141/files#r23503440>.
>